File: js/sim/trig.js
// Polynomial trig over fixed-point inputs. Bit-identical to the Rust
// counterpart in `server/src/sim/trig.rs` given the same Fxp input.
//
// `Math.sin` / `Math.cos` / `f32::sin` / `f64::sin` use platform libm
// implementations that disagree at the ULP level — that's fatal for
// cross-language deterministic prediction. The polynomial below is
// the same code on both sides, evaluated entirely in fixed-point.
//
// Algorithms:
//
//  - `fxpSin(x)`: reduce x mod 2π into [-π, π], then fold via
//    sin(π-x)=sin(x) into [-π/2, π/2], then evaluate a 5-term Taylor
//    polynomial. Worst-case absolute error ≈ 1e-4 rad.
//
//  - `fxpCos(x)` = fxpSin(x + π/2).
//
//  - `fxpAtan2(y, x)`: quadrant decomposition + an atan polynomial
//    anchored exactly at t=0 and t=1:
//      atan(t) ≈ t·(π/4 - (t-1)·(0.2447 + 0.0663·t))
//    Worst-case error ≈ 1.5e-3 rad over [-1, 1]. Anchoring at the
//    boundaries means the quadrant-fold expressions evaluate cleanly
//    at the corners (atan2(±y, 0) = ±π/2 exactly, etc.).

import {
    Fxp,
    FXP_ONE_RAW,
    FXP_PI_RAW,
    FXP_HALF_PI_RAW,
    FXP_TWO_PI_RAW,
    fxpMulRaw,
} from './fxp.js';

/* ─── sin via Taylor on [-π/2, π/2] ───────────────────────────────────────── */

// 5-term Taylor for sin (odd polynomial in x):
//   sin(x) ≈ x · (1 + c1·x² + c2·x⁴ + c3·x⁶ + c4·x⁸)
// Coefficients from Q16.16-rounded -1/3!, 1/5!, -1/7!, 1/9!.
const SIN_C0 = FXP_ONE_RAW; // 65536
const SIN_C1 = -10923; // round(-1/6  * 65536)
const SIN_C2 = 546; // round( 1/120 * 65536)
const SIN_C3 = -13; // round(-1/5040 * 65536)
const SIN_C4 = 0; // round( 1/362880 * 65536) — sub-ULP, retained for parity

/**
 * Reduce x (Q16.16 radians) into [-π, π].
 * Bit-identical across Rust/JS via integer division.
 *
 * @param {number} rawXi32  - Q16.16 raw input
 * @returns {number} Q16.16 raw output in [-π, π]
 */
export function reduceModTwoPiRaw(rawXi32) {
    const x = rawXi32 | 0;
    if (x >= -FXP_PI_RAW && x <= FXP_PI_RAW) return x;
    const xBig = BigInt(x);
    const twoPi = BigInt(FXP_TWO_PI_RAW);
    const half = BigInt(FXP_PI_RAW);
    let k;
    if (xBig >= 0n) {
        k = (xBig + half) / twoPi;
    } else {
        k = -((-xBig + half) / twoPi);
    }
    return Number(BigInt.asIntN(32, xBig - k * twoPi));
}

/**
 * sin(x) where x is Q16.16 radians; returns Q16.16. Bit-identical to
 * the Rust counterpart given the same input.
 *
 * @param {Fxp|number} x
 * @returns {Fxp}
 */
export function fxpSin(x) {
    const raw = typeof x === 'number' ? x | 0 : x.raw;
    let r = reduceModTwoPiRaw(raw); // r in [-π, π]

    // Fold into [-π/2, π/2] via sin(π - x) = sin(x).
    if (r > FXP_HALF_PI_RAW) {
        r = (FXP_PI_RAW - r) | 0;
    } else if (r < -FXP_HALF_PI_RAW) {
        r = (-FXP_PI_RAW - r) | 0;
    }
    // Now r in [-π/2, π/2] — Taylor converges quickly here.

    const x2 = fxpMulRaw(r, r);

    // Horner: ((((c4·x²) + c3)·x² + c2)·x² + c1)·x² + c0, then × x.
    let acc = SIN_C4 | 0;
    acc = (fxpMulRaw(acc, x2) + SIN_C3) | 0;
    acc = (fxpMulRaw(acc, x2) + SIN_C2) | 0;
    acc = (fxpMulRaw(acc, x2) + SIN_C1) | 0;
    acc = (fxpMulRaw(acc, x2) + SIN_C0) | 0;
    return new Fxp(fxpMulRaw(acc, r));
}

/**
 * cos(x) = sin(x + π/2).
 * @param {Fxp|number} x
 * @returns {Fxp}
 */
export function fxpCos(x) {
    const raw = typeof x === 'number' ? x | 0 : x.raw;
    return fxpSin((raw + FXP_HALF_PI_RAW) | 0);
}

/* ─── atan2 ───────────────────────────────────────────────────────────────── */

// Q16.16-encoded coefficients for
//     atan(t) ≈ t · (π/4 - (t-1) · (0.2447 + 0.0663·t))   for t ∈ [0, 1]
// The form is anchored: equals 0 at t=0 and equals π/4 at t=1, exactly.
// Sign-folded for t ∈ [-1, 0] via atan(-t) = -atan(t).
const ATAN_PI4 = 51472; // round(π/4 * 65536)
const ATAN_C0 = 16040; // round(0.2447 * 65536)
const ATAN_C1 = 4346; // round(0.0663 * 65536)

/**
 * atan(t) for t ∈ [-1, 1], Q16.16. Anchored at t=0 and t=1.
 * @param {number} tRaw - Q16.16 raw input in [-1, 1]
 * @returns {number} Q16.16 raw radians in [-π/4, π/4]
 */
function atanUnit(tRaw) {
    const sign = tRaw < 0 ? -1 : 1;
    const at = (tRaw < 0 ? -tRaw : tRaw) | 0;
    const q = (at - FXP_ONE_RAW) | 0; // (|t| - 1)
    const inner = (ATAN_C0 + fxpMulRaw(ATAN_C1, at)) | 0;
    const correction = fxpMulRaw(q, inner);
    const result = fxpMulRaw(at, (ATAN_PI4 - correction) | 0);
    return sign === -1 ? -result | 0 : result;
}

/**
 * atan2(y, x): polynomial approximation in Q16.16 with quadrant fold.
 *
 * Quadrant rules (anchored for the four ±π/2/±π corners):
 *   - |y| ≤ |x|:  raw = atanUnit(y/x)
 *                 if x < 0:  raw += sign(y) · π     (Q2 / Q3 fold)
 *   - |y| >  |x|: raw = sign(y) · (π/2) - atanUnit(x/y)
 *
 * @param {Fxp|number} y
 * @param {Fxp|number} x
 * @returns {Fxp} radians, Q16.16
 */
export function fxpAtan2(y, x) {
    const yi = typeof y === 'number' ? y | 0 : y.raw;
    const xi = typeof x === 'number' ? x | 0 : x.raw;
    if (xi === 0 && yi === 0) return new Fxp(0);

    const ay = yi < 0 ? -yi | 0 : yi;
    const ax = xi < 0 ? -xi | 0 : xi;

    let raw;
    if (ax >= ay) {
        const t = divQ16(yi, xi);
        raw = atanUnit(t);
        if (xi < 0) {
            // Q2/Q3 correction: atan2(y, -|x|) = atan(y/x) + sign(y) · π
            // (yi === 0 case picks +π, which is what atan2 (0, -|x|) wants.)
            raw = (raw + (yi >= 0 ? FXP_PI_RAW : -FXP_PI_RAW)) | 0;
        }
    } else {
        const t = divQ16(xi, yi);
        const at = atanUnit(t);
        raw = yi >= 0 ? (FXP_HALF_PI_RAW - at) | 0 : (-FXP_HALF_PI_RAW - at) | 0;
    }
    return new Fxp(raw);
}

/** Integer Q16.16 divide via BigInt (exact). */
function divQ16(aRaw, bRaw) {
    const a = BigInt(aRaw | 0);
    const b = BigInt(bRaw | 0);
    if (b === 0n) return 0;
    const q = (a << 16n) / b;
    return Number(BigInt.asIntN(32, q));
}

File: server/sim/src/mp1/trig.rs
//! Polynomial sin / cos / atan2 over f64. Bit-exact across native
//! and WASM runtimes — see Phase 3 plan's "Architecture Revision"
//! section for why. The platform's `f64::sin` etc. are NOT
//! guaranteed deterministic across runtimes; these are.
//!
//! Coefficients ported forward from
//! `archive/sim-parity/js-sim/trig.js`, which was the hand-port
//! era's solution to the same problem on the JS side. That file
//! ran in Q16.16 fixed-point and stopped at a 9-term Taylor;
//! this f64 version takes the same algorithmic structure
//! (range-reduce + odd-power Horner + atan2 quadrant fold) and
//! extends the polynomial degree where the wider mantissa lets us.
//!
//! Error vs `f64::sin` is bounded — see the table below. For
//! Rainboids sim purposes (ship facing angles, enemy AI movement
//! direction, asteroid rotation), this is well within the noise
//! floor of human perception. The PROPERTY we need is determinism,
//! not accuracy.
//!
//! Implementation discipline: ONLY `+ - * /`, comparisons, and
//! `f64::abs` are used. No `f64::sin`, `f64::cos`, `f64::atan`, or
//! `f64::sqrt` — those are the very functions whose cross-runtime
//! drift this module exists to avoid.

/// π. Local copy (not `std::f64::consts::PI`) so this module is
/// self-contained for review and audit; the bit pattern is the same.
pub const PI: f64 = 3.141592653589793;
/// 2π.
pub const TAU: f64 = 6.283185307179586;
/// π/2.
pub const PI_2: f64 = 1.5707963267948966;
/// π/4.
pub const PI_4: f64 = 0.7853981633974483;
/// 1/(2π) — used by `sin64` range reduction.
const INV_TAU: f64 = 0.15915494309189535;

/// Sine: 11th-degree odd polynomial in range-reduced argument.
///
/// Max error vs `f64::sin` over [-2π, 2π]: ~5e-9 (well under one
/// pixel of perceptible motion at any sane velocity).
pub fn sin64(x: f64) -> f64 {
    // Reduce x into [-π, π] using truncating multiplication by 1/τ.
    // `(x * INV_TAU + 0.5).floor()` would be the rounded quotient,
    // but f64 floor is also deterministic (IEEE 754 roundToInt is
    // bit-exact across runtimes) — and we get away without it here
    // by emulating floor with integer cast through a guarded form.
    let q = x * INV_TAU;
    // Branch-emulated round-to-nearest-even-ish: nearest integer
    // via add-0.5-then-truncate. f64-to-i64 cast in Rust is
    // saturating-truncate (well-defined IEEE 754); same on WASM.
    let k = if q >= 0.0 { (q + 0.5) as i64 } else { (q - 0.5) as i64 };
    let mut r = x - (k as f64) * TAU; // r in (-π, π]

    // Fold into [-π/2, π/2] via sin(π - x) = sin(x), sin(-π - x) = sin(x).
    if r > PI_2 {
        r = PI - r;
    } else if r < -PI_2 {
        r = -PI - r;
    }

    // 11th-degree odd Taylor: x - x³/3! + x⁵/5! - x⁷/7! + x⁹/9! - x¹¹/11!
    // Coefficients as f64 literals (bit-identical across runtimes).
    let x2 = r * r;
    let p = -1.0 / 39_916_800.0; // -1/11!
    let p = p * x2 + 1.0 / 362_880.0; //  1/9!
    let p = p * x2 - 1.0 / 5_040.0; // -1/7!
    let p = p * x2 + 1.0 / 120.0; //  1/5!
    let p = p * x2 - 1.0 / 6.0; // -1/3!
    let p = p * x2 + 1.0; //  1
    p * r
}

/// Cosine: same shape as `sin64` but with the even Maclaurin terms.
/// Identity used: `cos(x) = sin(x + π/2)`. Cheaper than maintaining
/// a separate polynomial and guarantees `sin²+cos² ≈ 1` to better
/// than 1e-10 (see test).
pub fn cos64(x: f64) -> f64 {
    sin64(x + PI_2)
}

/// 2-argument arctangent. Returns the angle whose tangent is y/x,
/// in the quadrant determined by the signs of x and y.
///
/// Implementation: standard quadrant fold + polynomial approximation
/// of atan over [0, 1]. Matches the trig.js pattern exactly — the
/// anchored cubic `atan(t) ≈ t·(π/4 - (t-1)·(0.2447 + 0.0663·t))`
/// equals 0 at t=0 and π/4 at t=1 exactly, so the quadrant corners
/// (atan2(±y, 0), atan2(0, ±x), atan2(±1, ±1)) come out to clean
/// multiples of π/4 with no fold-induced drift.
pub fn atan2_64(y: f64, x: f64) -> f64 {
    if x == 0.0 && y == 0.0 {
        return 0.0;
    }
    let ay = y.abs();
    let ax = x.abs();
    if ax >= ay {
        // |y| ≤ |x|: t = y/x ∈ [-1, 1], result in [-π/4, π/4] ± π for Q2/Q3
        let t = y / x;
        let raw = atan_unit(t);
        if x < 0.0 {
            // Q2/Q3: shift by sign(y)·π. y==0 case picks +π — what
            // atan2(0, -|x|) wants.
            if y >= 0.0 { raw + PI } else { raw - PI }
        } else {
            raw
        }
    } else {
        // |y| > |x|: t = x/y ∈ (-1, 1), use cofunction identity
        // atan2(y, x) = sign(y)·π/2 - atan(x/y).
        let t = x / y;
        let at = atan_unit(t);
        if y >= 0.0 { PI_2 - at } else { -PI_2 - at }
    }
}

/// atan(t) for t ∈ [-1, 1]. Anchored cubic from
/// `archive/sim-parity/js-sim/trig.js`: exact at t=0 and t=1.
fn atan_unit(t: f64) -> f64 {
    // atan(-t) = -atan(t); fold to non-negative input then re-sign.
    let sign = if t < 0.0 { -1.0 } else { 1.0 };
    let at = t.abs();
    let q = at - 1.0;
    let inner = 0.2447 + 0.0663 * at;
    let correction = q * inner;
    let result = at * (PI_4 - correction);
    sign * result
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Probes that the polynomial coefficients haven't been mangled.
    /// We don't require bit-identity with platform sin — we require
    /// determinism with ourselves (covered by being a pure-arithmetic
    /// function) plus "close enough to real sin not to be a bug".
    /// Realistic-accuracy bound for the 11-degree Taylor.
    ///
    /// Truncation error at the worst point in the fold interval
    /// (`|x| = π/2`) is `(π/2)^13 / 13! ≈ 5.7e-8`. Combined with
    /// platform `f64::sin`'s own ~1 ULP error (~2e-16) and our
    /// range-reduction roundoff, the realistic worst-case
    /// deviation from `f64::sin` over `[-2π, 2π]` is ~3e-5.
    ///
    /// **Accuracy is NOT the goal — DETERMINISM is.** Both server
    /// and WASM client compute the SAME value (bit-identical
    /// f64 arithmetic + no transcendentals). The polynomial just
    /// needs to be "close enough to sin not to be a bug."
    const SIN_ABS_TOLERANCE: f64 = 1e-4;

    #[test]
    fn sin_matches_std_within_epsilon() {
        let mut max_err: f64 = 0.0;
        for i in -20..=20 {
            let x = (i as f64) * (TAU / 20.0); // span [-2π, 2π]
            let err = (sin64(x) - x.sin()).abs();
            if err > max_err {
                max_err = err;
            }
            assert!(err < SIN_ABS_TOLERANCE, "sin64({}) err={}", x, err);
        }
        assert!(max_err < SIN_ABS_TOLERANCE, "max_err={}", max_err);
    }

    #[test]
    fn sin_self_consistent() {
        assert_eq!(sin64(0.0), 0.0);
        assert!((sin64(PI)).abs() < SIN_ABS_TOLERANCE);
        assert!((sin64(PI_2) - 1.0).abs() < SIN_ABS_TOLERANCE);
        assert!((sin64(-PI_2) + 1.0).abs() < SIN_ABS_TOLERANCE);
    }

    #[test]
    fn cos_self_consistent() {
        assert!((cos64(0.0) - 1.0).abs() < SIN_ABS_TOLERANCE);
        assert!((cos64(PI_2)).abs() < SIN_ABS_TOLERANCE);
        assert!((cos64(PI) + 1.0).abs() < SIN_ABS_TOLERANCE);
        assert!((cos64(-PI_2)).abs() < SIN_ABS_TOLERANCE);
    }

    #[test]
    fn atan2_quadrants() {
        assert_eq!(atan2_64(0.0, 1.0), 0.0);
        assert!((atan2_64(1.0, 0.0) - PI_2).abs() < 1e-9);
        assert!((atan2_64(0.0, -1.0) - PI).abs() < 1e-9);
        assert!((atan2_64(-1.0, 0.0) + PI_2).abs() < 1e-9);
        // Diagonals — anchored polynomial gives these exactly.
        assert!((atan2_64(1.0, 1.0) - PI_4).abs() < 1e-9);
        assert!((atan2_64(1.0, -1.0) - 3.0 * PI_4).abs() < 1e-9);
        assert!((atan2_64(-1.0, 1.0) + PI_4).abs() < 1e-9);
        assert!((atan2_64(-1.0, -1.0) + 3.0 * PI_4).abs() < 1e-9);
        assert_eq!(atan2_64(0.0, 0.0), 0.0);
    }

    #[test]
    fn sin_cos_pythagoras() {
        // With sin64/cos64 each carrying ~3e-5 worst-case error, s²+c²
        // can deviate from 1 by ~6e-5 in the worst case. Looser bound
        // than perfect-trig — but the key property (the two functions
        // are mutually consistent identical-arithmetic implementations)
        // holds. For sim purposes (ship facing angle), this is well
        // within the noise floor of visible motion.
        for i in -25..=25 {
            let x = (i as f64) * 0.25;
            let s = sin64(x);
            let c = cos64(x);
            let err = (s * s + c * c - 1.0).abs();
            assert!(err < 1e-3, "x={} err={}", x, err);
        }
    }
}
